docs/tutorials/client-server.md

```diff
@@ -0,0 +1,230 @@
+
+- Add a [`Zip`](xref:Bonsai.Reactive.Zip) operator to the [`Index`](xref:Bonsai.Expressions.IndexBuilder) operator and connect the `byte[]` `Buffer` as the second input.
+
+:::workflow
+![Address key-value pair](~/workflows/address-kvp.bonsai)
+:::
+
+Every time the [`Router`](xref:Bonsai.ZeroMQ.Router) receives a message, the [`Zip`](xref:Bonsai.Reactive.Zip) will create a `Tuple` that can be thought of as a key-value pair, with the unique `byte` address of the client as the key, and the full `byte[]` address used by ZeroMQ for routing as the value. 
+
+- Add a [`DistinctBy`](xref:Bonsai.Reactive.DistinctBy) operator after the [`Zip`](xref:Bonsai.Reactive.Zip) and set the `KeySelector` property to the `byte` value (`Item1`).
+
+:::workflow
+![Unique key-value pair](~/workflows/unique-kvp.bonsai)
+:::
+
+The [`DistinctBy`](xref:Bonsai.Reactive.DistinctBy) operator filters the output of [`Zip`](xref:Bonsai.Reactive.Zip) according to the unique `byte` value and produces a sequence containing only the distinct – or ‘new’ – values produced by [`Zip`](xref:Bonsai.Reactive.Zip). The output of [`DistinctBy`](xref:Bonsai.Reactive.DistinctBy) will therefore effectively be a sequence of unique client addresses corresponding to each connected client. We also need to store these unique values and make them available to other parts of the Bonsai workflow. 
+
+- Add a [`ReplaySubject`](xref:Bonsai.Reactive.ReplaySubject) operator after [`DistinctBy`](xref:Bonsai.Reactive.DistinctBy) and name it ‘ClientAddresses’. 
+
+:::workflow
+![Address ReplaySubject](~/workflows/address-replay-subject.bonsai)
+:::
+
+A [`ReplaySubject`](xref:Bonsai.Reactive.ReplaySubject) has the useful feature that it stores its input sequence and replays those values to any current or future subscribers. The effect in this case is that anything that subscribes to `ClientAddresses` will receive all the unique client addresses encountered by the server so far.
+
+## Server --> client communication
+Eventually, we will use these unique client addresses to route server messages back to specific client. For now, we'll implement a more basic approach where the server just sends messages back to the client that originally sent them. The Bonsai ZeroMQ library provides a convenient operator for this task in the form of [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse). 
+
+- Add a [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse) operator after the [`Router`](xref:Bonsai.ZeroMQ.Router) in a separate branch.
+- Inside (double-click on [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse)) add a [`String`](xref:Bonsai.Expressions.StringProperty) operator with a generic response value like `ServerResponse` after the `Source` operator. 
```

Comment:
```suggestion
- Double-click on the [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse) node to open up its internal workflow.
- Insert a [`String`](xref:Bonsai.Expressions.StringProperty) operator after the workflow `Source`, with a generic response value like `ServerResponse`. 
```

From other tutorials, we tend to describe the step of opening up a nested workflow separately from what we do inside.